src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,436 @@
+        """
+
+        if workspace_id is not None and workspace_name is not None:
+            raise ValueError("Specify either 'workspace_id' or 'workspace_name'!")
+        if dataset_id is not None and dataset_name is not None:
+            raise ValueError("Specify either 'dataset_id' or 'dataset_name'!")
+
+        self.workspace_id = workspace_id
+        self.workspace_name = workspace_name
+        self.dataset_id = dataset_id
+        self.dataset_name = dataset_name
+
+        # Set access parameters
+        self.client = client
+
+        self.max_minutes_after_last_refresh = max_minutes_after_last_refresh
+        self.min_refresh_time_utc = datetime.now(utc) - timedelta(
+            minutes=self.max_minutes_after_last_refresh
+        )
+        self.timeout_in_seconds = timeout_in_seconds
+        self.local_timezone_name = local_timezone_name
+        self.ignore_errors = ignore_errors
+        self.api_header = None
+        self.expire_time = 0
+
+        self.last_status = None
+        self.last_exception = None
+        self.last_refresh_utc = None
+        self.last_duration = 0
+
+    def _raise_error(self, message: str):
+        if self.ignore_errors:
+            print(message)
+        else:
+            raise Exception(message)
+
+    def _raise_api_error(self, message: str, api_call: requests.Response):
+        print(api_call.text)
+        self._raise_error(
+            message + f" Response: {api_call.status_code} {api_call.reason}"
+        )
+
+    def _get_access_token(self) -> bool:
+        """
+        Acquires an access token to connect to PowerBI.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises Exception: if failed and ignore_errors==False
+        """
+
+        if self.expire_time != 0:
+            if time.time() < self.expire_time:
+                return True
+            print("Renewing access token...")
+
+        # Prepare URLs (note: we fetch only the latest refresh record, i.e. top=1)
+        authority_url = f"https://login.microsoftonline.com/{self.client.tenant_id}/"
+        scope = ["https://analysis.windows.net/powerbi/api/.default"]
+        self.powerbi_url = "https://api.powerbi.com/v1.0/myorg/"
+
+        # Use MSAL to get an access token
+        app = msal.ConfidentialClientApplication(
+            self.client.client_id,
+            authority=authority_url,
+            client_credential=self.client.client_secret,
+        )
+        result = app.acquire_token_for_client(scopes=scope)
+        if "access_token" not in result:
+            print(result)
+            self._raise_error("Failed to acquire token!")
+            return False
+
+        access_token = result["access_token"]
+        if "expires_in" in result:
+            self.expire_time = time.time() + result["expires_in"] - 3
+        else:
+            self.expire_time = time.time() + (5 * 60) - 3
```

Comment:
I don't like magic numbers in code. Could these be added as a constant with a defining name.

Reply:
Agreed. Fixed.